pkg/planner/core/rule_predicate_simplification.go

```diff
@@ -396,7 +412,7 @@ func processCondition(sctx base.PlanContext, condition expression.Expression) (e
 		condition, applied = shortCircuitANDORLogicalConstants(sctx, condition, false)
 	}
 
-	if applied {
+	if applied && expression.MaybeOverOptimized4PlanCacheForMultiExpression(sctx.GetExprCtx(), condition) {
```

Comment:
When it comes to here, the `condition` may already be changed, I think we need to do this check before `shortCircuitANDORLogicalConstants`.